### ISSUE TITLE
Support test_INa_IV_curves protocol

### ISSUE THREAD

**Issue by jonc125** [ModellingWebLab/weblab-fc#99]
Part of #23, following from #95 and #96 and #97 .

See https://chaste.cs.ox.ac.uk/trac/browser/chaste/projects/FunctionalCuration/test/TestOptionalVariables.hpp for tests of this protocol in the C++ code. We'll need to create equivalent testing in Python.

- [x] Implement `DeclareVariable` (for `var ... units ... = ...`) #154
- [x] Implement units conversion rules (`convert ... to ... by ...`; `UnitsConversion` action class) #156 

Up to here will allow testing on models that have `oxmeta:extracellular_sodium_concentration`, `oxmeta:cytosolic_sodium_concentration`, `oxmeta:cytosolic_potassium_concentration`, and `oxmeta:cytosolic_calcium_concentration`. Make sure current units conversions are needed!

Once these steps are done, `test_ical_proto.py` should also work. https://chaste.cs.ox.ac.uk/trac/browser/chaste/projects/FunctionalCuration/test/TestIcalProtocol.hpp gives ideas for a shorter version of this test.

- [x] Implement `OptionalVariable` handling (for `optional ...`) for clamped inputs. #157 

- [x] Once all the above is done, we may find that `test_clamping.py:test_clamp_to_data_file` just works. If not, open a separate issue for fixing what parts of it still need implementing.

**Comment by MichaelClerx** [comment]
Do we have a strategy for generating names for local (``var``) variables in error messages? Usually we'd use rdf terms but they don't have to have any.

**Comment by jonc125** [comment]
I'd guess report just the variable name? Since that's the name given in the protocol so will be what the user expects?

**Comment by MichaelClerx** [comment]
> I'd guess report just the variable name? Since that's the name given in the protocol so will be what the user expects?

Well, you might not remember what the local name for it was at the point that you're creating the message, because its name in the model will be e.g. `protocol__x_3` instead of `x`

**Comment by MichaelClerx** [comment]
@jonc125 what do I do with ``var`` variables from nested/imported protocols? 

I'm guessing we need some sort of mechanism so that each protocol has its own namespacing / scope for these?

**Comment by jonc125** [comment]
Ah, I see the problem. So perhaps what we want is an attribute on `VariableDummy` objects that we can set to give an 'informative name' for the variable? And the `DeclareVariable` class might need to remember the protocol file path it came from, so we can include the protocol file name and local var name in this informative name? (Probably not the full protocol path; typically all files will be in the same folder and have distinct names, and we don't want to leak filesystem paths from the server if we don't have to!)

**Comment by MichaelClerx** [comment]
Ok, will think about that.

More about nested vars in general: If we merge/flatten all the protocols before doing anything, how do I prevent a user e.g. creating a variable in an outer protocol, and then defining it in a nested one?

**Comment by jonc125** [comment]
You don't, that's OK. Might not be entirely sensible, but it should still work. For instance you might design a common.txt protocol with some declarations etc., but want to define a variable differently in two protocols importing that.

**Comment by MichaelClerx** [comment]
Should I detect clashes when they both define a var with name x?

**Comment by jonc125** [comment]
> Should I detect clashes when they both define a var with name x?

Yes, that should be an error.
